src/RedisIPCMessenger.ts

```diff
@@ -0,0 +1,119 @@
+import Redis from 'ioredis';
+import IPCMessenger, {
+  Room,
+  Instance,
+  MessageCallback,
+  MessageTypes,
+  Message,
+} from './IPCMessenger';
+
+type ConstructorParams = {
+  room: Room;
+  instance: Instance;
+  callback: MessageCallback,
+  redis?: Redis.RedisOptions,
+  expireTime?: number,
+  refreshTime?: number,
+}
+
+class RedisIPCMessenger implements IPCMessenger {
+  private readonly room: Room;
+  private readonly instance: Instance;
+  private readonly callback: MessageCallback;
+  private readonly publisher: Redis.Redis;
+  private readonly subscriber: Redis.Redis;
+  private readonly expireTime: number;
+  private readonly refreshTime: number;
+  private hasJoined: boolean;
+
+  constructor(params: ConstructorParams) {
+    this.validateInput(params);
+
+    this.room = params.room;
+    this.instance = params.instance;
+    this.callback = params.callback;
+
+    this.publisher = new Redis(params.redis);
+    this.subscriber = new Redis(params.redis);
+    this.expireTime = params.expireTime ?? 30000;
+    this.refreshTime = params.refreshTime ?? 10000;
+    this.hasJoined = false;
+  }
+
+  async join(): Promise<void> {
+    if (this.hasJoined) {
+      throw new Error('Can\'t join twice!');
+    }
+
+    const refreshKeyLoop = async () => {
```

Comment:
It's now in a separate method, the loop itself is still inlined, but now there is an indication that the method is just starting the loop, otherwise you would end up awaiting loop which looks more confusing.

Reply:
I don't understand - why don't we flatten it into one method? The effect  will be the same:

```js
  private startRefreshKeyLoop = async (room: Room) => {
    await this.publisher.set(`${room}:${this.instance}`, '', 'EX', this.expireTime / 1000);
    setTimeout(this.startRefreshKeyLoop, this.refreshTime);
  };
```

Reply:
The name can changed to `loopKeyRefresh` or something.

Reply:
I don't mind that

Reply:
Uh? I still don't understand. Making code more readable is great as long as you don't introduce actual inefficiencies. In this case `loop` can be used as a verb - "go loop the key refresh procedure". We're not starting anything.

Reply:
It's the exact same thing pefromance-wise, but I don't find HUGE diff between the two anyways. Will change it.

Reply:
Well, almost the same thing - the nested approach would create a new function object every time the outer function is called.